### PR TITLE
fix: ensure (a?.b)() has proper this

### LINKED ISSUE #10802: Optional Chaining does not propagate receiver when parenthesized
## Bug Report

**Current Behavior**

Optional Calls allows you to parenthesize the expression to break the optional-ness of the chain. For instance, `(foo?.bar).baz` requires that an object be returned to access `.baz`. This also applies to call expressions: `(foo?.bar)()`.

The problem is, that `(foo?.bar)()` must propagate `foo` as the `this` receiver to the call expression. We currently do not do this.

**Input Code**

[repl](https://babeljs.io/repl/#?browsers=Chrome%2078&build=&builtIns=usage&spec=false&loose=false&code_lz=MYewdgzgLgBAZiEMC8MDeMBGBDATgCgEp0ZcBTKAV1zBigAsBLCGAXzYG4AoL0SEADZkAdNggQyuKPgRJk8mDMQB-YTlyEihDkA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=&prettier=false&targets=&version=7.7.4&externalPlugins=%40babel%2Fplugin-proposal-optional-chaining%407.6.0)

```js
const foo = { bar() { return this } };

console.assert(foo === (foo?.bar)());
```

**Expected behavior/code**
The `foo` object should be used as the receiver during the `foo.bar` method call.

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**
- Filename: `babel.config.js`

```js
{
  "plugins": [ "@babel/plugin-proposal-optional-chaining" ]
}
```

**Possible Solution**
We need to detect parenthesized call expressions, and use `.call` to properly propagate the receiver.

**Additional context/Screenshots**

This is likely more difficult good-first-issue. I'm still prototyping a fix to see whether we should recommend it to a first time contributor.

### PR DESCRIPTION
**Pull request by JLHwung**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #10802, closes #11327 
| Patch: Bug Fix?          | Yes
| Tests Added + Pass?      | Yes
| License                  | MIT

It has been out there for quite a while and the fix seems more tricky than previously proposed in #11327 and #10802.

This PR supersedes #11552 .
 
Todo
- [x] `(a?.#m)()`
- [x] `parserOpts.createParenthesizedExpression`

### CODE REVIEW

File: packages/babel-helper-member-expression-to-functions/src/index.js [hunk 1]
```diff
@@ -167,13 +167,19 @@ const handle = {
       const parentIsOptionalCall = parentPath.isOptionalCallExpression({
         callee: node,
       });
+      const isParenthesizedMemberCall =
+        parentPath.isCallExpression({ callee: node }) &&
+        node.extra?.parenthesized;
```
Comment (JLHwung) [inline packages/babel-helper-member-expression-to-functions/src/index.js:172]: The current approach does not support `parserOpts.createParenthesizedExpressions`. When developing I realize that the whole transform does not take `ParenthesizedExpressions` into account, I suggest we address it in a separate PR before this one gets too big to review.

File: packages/babel-plugin-proposal-optional-chaining/src/index.js [hunk 2]
```diff
@@ -23,14 +23,23 @@ export default declare((api, options) => {
...
         let isDeleteOperation = false;
         const optionals = [];
 
         let optionalPath = path;
         while (
           optionalPath.isOptionalMemberExpression() ||
-          optionalPath.isOptionalCallExpression()
+          optionalPath.isOptionalCallExpression() ||
+          optionalPath.isParenthesizedExpression() ||
+          optionalPath.isTSNonNullExpression()
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-optional-chaining/src/index.js:42]: By doing so we also fix cases like `a!!?.b` are not transformed currently.

File: packages/babel-helper-create-class-features-plugin/src/fields.js [hunk 3]
```diff
@@ -323,6 +330,13 @@ const privateNameHandlerLoose = {
     });
   },
 
+  boundGet(member) {
+    return t.callExpression(
+      t.memberExpression(this.get(member), t.identifier("bind")),
+      [member.node.object],
```
Comment (jridgewell) [inline packages/babel-helper-create-class-features-plugin/src/fields.js:336]: Don't we need to clone?

File: packages/babel-helper-create-class-features-plugin/src/fields.js [hunk 4]
```diff
@@ -241,6 +241,13 @@ const privateNameHandlerSpec = {
     ]);
   },
 
+  boundGet(member) {
+    return t.callExpression(
+      t.memberExpression(this.get(member), t.identifier("bind")),
```
Comment (jridgewell) [inline packages/babel-helper-create-class-features-plugin/src/fields.js:246]: Don't we need to memoize the member?

File: packages/babel-plugin-proposal-optional-chaining/src/index.js [hunk 5]
```diff
@@ -23,14 +23,23 @@ export default declare((api, options) => {
...
       "OptionalCallExpression|OptionalMemberExpression"(path) {
-        const { parentPath, scope } = path;
+        const { scope } = path;
+        let parentPath;
+        // unwrap parenthesis around parent path
+        for (
+          { parentPath } = path;
+          parentPath.type === "ParenthesizedExpression";
+          { parentPath } = parentPath
+        );
```
Comment (jridgewell) [inline packages/babel-plugin-proposal-optional-chaining/src/index.js:33]: Maybe we should call it something other than `parentPath`?

```suggestion
        const parentPath = path.findParent(p => !p.isParenthesizedExpression());
```

File: packages/babel-plugin-proposal-optional-chaining/src/index.js [hunk 6]
```diff
@@ -113,7 +120,36 @@ export default declare((api, options) => {
...
+              // it must be unwrapped from parenthesized expression.
+              replacementPath.parentPath !== parentPath) &&
+            parentPath.isCallExpression()
+          ) {
+            // `(a?.b)()` to `(a == null ? undefined : a.b.bind(a))()`
+            const { object } = replacement;
+            let baseRef;
+            if (!loose) {
+              // memoize the context object in non-loose mode
+              // `(a?.b.c)()` to `(a == null ? undefined : (_a$b = a.b).c.bind(_a$b))()`
```
Comment (jridgewell) [inline packages/babel-plugin-proposal-optional-chaining/src/index.js:138]: Do we need to check if there's a call expression in the path? `(o?.Foo().m)();` will double call, I think.

File: packages/babel-plugin-proposal-optional-chaining/src/index.js (hunk 2 above)
Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-optional-chaining/src/index.js:36]: @oliverdunk Should we also check for this in your PR?

Comment (oliverdunk) [reply packages/babel-plugin-proposal-optional-chaining/src/index.js:36]: Makes sense, I'll add that now 👍

**Pull request merged**